### PR TITLE
Add warning diagnostic when using V2 api

### PR DESCRIPTION
**Pull request by microbit-robert**
- [ ] Translation sync required before merge and use (or at least addition of new strings in all simplified language files in English for now). Keeping in draft until this is done.

### CODE REVIEW

File: packages/pyright-internal/src/analyzer/binder.ts [hunk 1]
```diff
@@ -386,6 +387,19 @@ export class Binder extends ParseTreeWalker {
...
+            this._addDiagnostic(
+                this._fileInfo.diagnosticRuleSet.reportMicrobitV2ApiUse,
+                DiagnosticRule.reportMicrobitV2ApiUse,
+                Localizer.Diagnostic.microbitV2ModuleUse().format({
+                    moduleName: importResult.importName,
+                    device,
+                }),
+                node
+            )
+        );
```
Comment (microbit-robert) [inline packages/pyright-internal/src/analyzer/binder.ts:401]: This is setup with a callback so we can track micro:bit specific diagnostics in one file, `microbitUtils.ts`. Useful if we only want to warn on the first use of a V2 API call in the future. We can't directly pass in `this_.addDiagnostic` as this causes errors.

File: packages/pyright-internal/src/analyzer/microbitUtils.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+function usesMicrobitV2Api(moduleName: string, name?: string) {
+    return (
+        ['log', 'microbit.microphone', 'microbit.speaker', 'power'].includes(moduleName) ||
+        (moduleName === 'microbit' &&
+            ['run_every', 'set_volume', 'Sound', 'pin_logo', 'pin_speaker'].includes(name ?? '')) ||
+        (moduleName === 'microbit.audio' && name === 'SoundEffect') ||
+        (moduleName === 'neopixel' && ['fill', 'write'].includes(name ?? ''))
+    );
```
Comment (microbit-robert) [inline packages/pyright-internal/src/analyzer/microbitUtils.ts]: Hardcoded, but only in one place. Are any missing?

File: packages/pyright-internal/src/localization/simplified.nls.en-us.json [hunk 3]
```diff
@@ -58,6 +58,9 @@
         "memberSet": "Cannot assign member \"{name}\" for type \"{type}\"",
         "moduleNotCallable": "Module is not callable",
         "moduleUnknownMember": "\"{name}\" is not a known member of module \"{module}\"",
+        "microbitV2ClassMethodUse": "The \"{methodName}\" method on the \"{className}\" class cannot be used with a {device}",
+        "microbitV2ModuleMemberUse": "\"{name}\" from the \"{moduleName}\" module cannot be used with a {device}",
+        "microbitV2ModuleUse": "\"{moduleName}\" cannot be used with a {device}",
```
Comment (microbit-robert) [inline packages/pyright-internal/src/localization/simplified.nls.en-us.json:63]: Highlighting for review as these will be tweaked based on prior discussion.

**Comment by microbit-matt-hillsdon** [comment]
I've tried for an alternative on https://github.com/microbit-foundation/pyright/pull/35

Mainly focussed on the layer we do this at but I've also made some incidental changes to the error code/text.

**Comment by microbit-matt-hillsdon** [comment]
Merged the alternative on https://github.com/microbit-foundation/pyright/pull/35
Closing this one

**Pull request closed without merging**